### PR TITLE
RFC Merge RPM 4.13 OS/2 code changes

### PR DESCRIPTION
**Pull request by ydario**
I have committed the generic patches to RPM source tree. Now that I know git better, I applied my patches to master.

### CODE REVIEW

File: fileattrs/python.attr [hunk 1]
```diff
@@ -1,4 +1,4 @@
 %__python_provides	%{_rpmconfigdir}/pythondeps.sh --provides
 %__python_requires	%{_rpmconfigdir}/pythondeps.sh --requires
-%__python_path	^((/usr/lib(64)?/python[[:digit:]]\\.[[:digit:]]/.*\\.(py[oc]?|so))|(%{_bindir}/python[[:digit:]]\\.[[:digit:]]))$
+%__python_path	^((%{_prefix}/lib(64)?/python[[:digit:]]\\.[[:digit:]]/.*\\.(py[oc]?|so|pyd))|(%{_bindir}/python[[:digit:]]\\.[[:digit:]](\\.exe)?))$
```
Comment (ignatenkobrain) [inline fileattrs/python.attr]: pyd? what's that?

Comment (ydario) [inline fileattrs/python.attr]: it is the dll extension used for python code.

**Comment by ignatenkobrain** [comment]
Let's try to reopen and see if CI passes.

**Comment by ydario** [comment]
No, there is other work which is not ready for merging, see https://github.com/bitwiseworks/rpm/tree/master-os2-dev; but these patches are addressing major OS/2 differences.

**Comment by pmatilai** [comment]
Sorry but there's just no way we can justify this amount of added code + complexity = maintenance burden for a long dead OS.

**Comment by ydario** [comment]
OS/2 is currently mantained and developed by ArcaNoae LLC, they released ArcaOS 5.0 and many updates to drivers and applications:

https://www.arcanoae.com/tag/arcaos/

Also a lot of work has been done in porting open source applications

http://trac.netlabs.org/rpm
http://trac.netlabs.org/ports

and in general http://trac.netlabs.org/

Merging RPM is one fundamental step in our effort.

**Comment by pmatilai** [comment]
An OS that is niche to the point that I haven't even heard of it is hardly more justifiable than a dead one. And it's not just a niche OS or a dead OS but it's also a non-POSIX, decidedly non-unix OS, so it's quite a leap we're talking about here.

And we've just been considering bumping the system requirement to POSIX.1-2008 (from 2001) to make certain things simpler/possible.

That said, there are a number of things in this patch that could be merged if done right. Some examples:
* The IMA plugin should not be built at all if <sys/xattr.h> and lxsetattr() are not present on the system. That check belongs to configure.ac + Makefiles and has nothing at all to do with OS/2, it's plain correctness.
* Similarly for ctermid(), test for the existence of the function, not OS.
* build/pack.c already includes <sys/wait.h>. If the order is wrong then change that instead of yet more ifdefs
* __python_path hardcoding /usr/lib is wrong anyway, but so is %{prefix} because rpm's prefix and python's prefix don't necessarily have anything to do with each other, it should discover the python path by some other means
* The added error check in doScript() for child == -1 would be correct for fork() too. I also wonder why is spawnvp() used there if everything else seems to be happy with fork()?
* Line- and path-separator issues could be handled centrally (configure.ac / system.h macro or such) without making it any worse for anybody

Quite possibly something else too, I didn't look *that* carefully. In general, the way to add portability is not to sprinkle ifdef's but add abstraction.

tools/os2deps.c is missing in that patch so no idea what it does, but the invocation through a script helper looks a bit strange, and this magic pattern looks overly broad, I'd expect there to be some further identification in that string:
+%__os2_magic           ^(32|64)-bit.*$

In any case it looks like that's dealing with os2 *executables* so a better name would be something less broad than "os2".

**Comment by pmatilai** [comment]
See eg commit 014246b0502e30a20a2efef6fdf5c177ccba3ece, I fixed that because it was an actual bug in the rpm codebase. That's one less thing for you to patch around, and rpm was actually improved as a result.

**Comment by glebfm** [comment]
> And we've just been considering bumping the system requirement to POSIX.1-2008 (from 2001) to make certain things simpler/possible.
I think we actually already did it with aeb58f6.

**Comment by pmatilai** [comment]
Ugh, silently introducing new (major) dependencies is not good :-/

**Comment by ydario** [comment]
While OS/2 itself is not a POSIX system, our build system is based on gcc+libc; the libc code is written to make a UNIX style environment available on OS/2 too.

**Comment by pmatilai** [comment]
I explained a whole pile of things that you could do to help your cause in  https://github.com/rpm-software-management/rpm/pull/260#issuecomment-339914006, how about starting with those?

**Pull request closed without merging**